File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double solve(double x, int n) {
        if(n == 0) return 1.0;
        double half = solve(x,n/2);
        if(n%2==0) return half*half;
        else return half*half*x;
    }

    double myPow(double x, int n) {
        double ans = 1;
        long long N = n;
        if(N<0) {
            N = -N;
            x = 1/x;
        }
        return solve(x,N);
    }
};

File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, int n) {
        long e = n;
        if(e<0) {
            x = 1/x;
            e = -e;
        }
        double a = 1;
        while(e>0) {
            if(e%2 == 1) {
                a *= x;

            }
            x *= x;
            e /= 2;
        }
        return a;
    }
}